### ISSUE TITLE
Implement prefix-lm as in the T5 paper

### ISSUE THREAD

**Issue by ibeltagy** [bigscience-workshop/Megatron-DeepSpeed#21]
AFAIU, the current implementation ~~uses the first 50% as the prefix~~ doesn't support prefix-lm.  The T5 paper samples the prefix length randomly from 0 to max_sequence_lengh (which is 2048 in our case)

**Comment by thomasw21** [comment]
Shouldn't the prefix-lm be on the size of the context?

**Comment by ontocord** [comment]
Happy to help w/ this if you point me to the right files.  @thomasw21 I can help you with this if you want. Lmk

**Comment by ibeltagy** [comment]
@thomasw21, I just saw your question now. 

Check this figure from the T5 paper 
![image](https://user-images.githubusercontent.com/2287797/127929740-e337fb0d-3ab7-4557-a25e-d1e16942d2b3.png)

With autoregressive LM, the mask is "Causal" (the one in the middle). 
With prefix-lm, the mask is "Causal with prefix" (the one on the right). 
The question is about the length of the prefix. The T5 implementation randomly samples a different prefix length between 0 to max (max is the sequence length) for each batch. 


@ontocord, 
As far as I can tell, the entry point is the 2 calls for the function `get_ltor_masks_and_position_ids` here 
https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/pretrain_gpt.py


CCing @TevenLeScao just as an FYI because he tried to implement it before in a different codebase.

**Comment by thomasw21** [comment]
Ah yes sorry I meant, why only randomly sampled? To my understanding:
 - for unlabeled data, you don't really have a choice, so random prefix makes sense.
 - for prompts we generated through the hackathon, we probably want to set the prefix to the context, ie everything before "|||" in jinja no?

**Comment by ibeltagy** [comment]
That's fine, we can implement one function that takes the prefix length as input or samples a random length if the input is None. In `pretrain_gpt.py`, we will call the random version.

**Comment by ontocord** [comment]
Thanks @thomasw21 and @ibeltagy. I'll take a look at https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/pretrain_gpt.py

**Comment by ontocord** [comment]
Unless I am mistaken, is it just a matter of setting

**Comment by ontocord** [comment]
Btw, are we trying to do something like UniLM? https://arxiv.org/pdf/1905.03197.pdf

**Comment by thomasw21** [comment]
> Unless I am mistaken, is it just a matter of setting attention_mask[:,:,:prefix_len,:] = 1.0 after this line?
> https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/781676b59d28c5f4b6fd155fcbf15026991d2187/megatron/utils.py#L162
> 
> 
> Will we be requiring some sort of mask-span for words like in T5 in the prefix portion?

I'm not 100% sure but essentially multiple samples are batch-up together in a single sequence and seperated by `eod` (End of document) token. So this should be slightly more tricky that doing this. But I can help.

Also we're planing to have an experiment running by the end of the week, would this work for you? If not I can help/take over.

**Comment by ontocord** [comment]
def get_ltor_masks_and_position_ids(data,
                                    eod_token,
                                    reset_position_ids,
                                    reset_attention_mask,
                                    eod_mask_loss,
                                    prefix_len=None):
    """Build masks and position id for left to right model."""

    # Extract batch size and sequence length.
    micro_batch_size, seq_length = data.size()
    if prefix_len is not None:
        if type(prefix_len) is int:
            prefix_len = [prefix_len]*micro_batch_size
        else len(prefix_len) != micro_batch_size:
            #prefix_len is some sequence indexable by [], which could be a list, or a tensor, etc.
            raise RuntimeError("prefix_len must be the same batch size as data")
            
    # Attention mask (lower triangular).
    if reset_attention_mask:
        att_mask_batch = micro_batch_size
    else:
        att_mask_batch = 1
    attention_mask = torch.tril(torch.ones(
        (att_mask_batch, seq_length, seq_length), device=data.device)).view(
            att_mask_batch, 1, seq_length, seq_length)

    # Loss mask.
    loss_mask = torch.ones(data.size(), dtype=torch.float, device=data.device)
    if eod_mask_loss:
        loss_mask[data == eod_token] = 0.0

    # Position ids.
    position_ids = torch.arange(seq_length, dtype=torch.long,
                                device=data.device)
    position_ids = position_ids.unsqueeze(0).expand_as(data)
    # We need to clone as the ids will be modifed based on batch index.
    if reset_position_ids:
        position_ids = position_ids.clone()
    if reset_position_ids or reset_attention_mask:
        # Loop through the batches:
        for b in range(micro_batch_size):

            # Find indecies where EOD token is.
            eod_index = position_ids[b, data[b] == eod_token]

            # Detach indecies from positions if going to modify positions.
            if reset_position_ids:
                eod_index = eod_index.clone()

            # Loop through EOD indecies:
            prev_index = 0
            for j in range(eod_index.size()[0]):
                i = eod_index[j]
                # Mask attention loss.
                if reset_attention_mask:
                    attention_mask[b, 0, (i + 1):, :(i + 1)] = 0
                    if prefix_len and prefix_len[b] < 0: 
                        attention_mask[b, 0, math.randInt(2, i-2): :] = 1.0 
                        # we assume that we have no examples with less than 3 tokens. we want prefix to be random num between 2 and one token before the eod token.
                    elif prefix_len and prefix_len[b] > 0: 
                        # it doesn't make sense to have a prefix == 1 because that is just auto-regression, so we limit to 2 to
                        attention_mask[b, 0, min(i-1, max(prefix_len[b], 2)): :] = 1.0
                                                
                # Reset positions.
                if reset_position_ids:
                    position_ids[b, (i + 1):] -= (i + 1 - prev_index)
                    prev_index = i + 1
                    

    # Convert attention mask to binary:
    attention_mask = (attention_mask < 0.5)

    return attention_mask, loss_mask, position_ids

**Comment by ontocord** [comment]
Lmk what you think of this @thomasw21 and @ibeltagy. It accounts for a batch with lots of different eod.

**Comment by ontocord** [comment]
Question: what happens to the positional after the prefix? Wouldn't this condition the model to process only where the right handside after the prexi starts with some positon > 1? But I'm not really sure because we are using rotary position right?

**Comment by sbmaruf** [comment]
Here's the details of all the architectures that `t5` model considers, https://docs.google.com/document/d/110tlidAcpiNteKnA27tR5KPS_VahNqYKqCeJlu1MWww/edit#
See `Architecture config mapping and Analysis on T5 Table 2`.

@ontocord @thomasw21  If it helps, here's the entry point of the prefix-lm [implementation](https://github.com/tensorflow/mesh/blob/master/mesh_tensorflow/transformer/transformer.py#L750) in mesh transformers. `t5` used this implementation.

**Comment by ontocord** [comment]
I posted this in our slack channel and reposting here so we have it in this thread:

I saw this in the code refering some sort of upper triangle optimization in fused_softmax.py. I couldn't figure out what the code actually does since it uses an external package (scaled_upper_triang_masked_softmax_cuda), but if it assumes a casual masking strategy, wouldn't this be messed up if we use a prefix which is not causual in the prefix part. Or maybe we don't do prediction for the prefix, so it doesn't matter?  https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/dc4e0cbad094043e3cab3bbcbcbc3080ea8f7363/megatron/model/fused_softmax.py#L24  Sorry if this question doesn't make sense as this is the first time I'm jumping into megatron training code and not pre-processing.

**Comment by thomasw21** [comment]
@ontocord that's a custom cuda kernel, it's suposed to do the same thing as the code just below it. Basically you can write your own to have improved performance. You can find the code in `*.cu` files if you're interested. In our case let's not do this as this might just take a lot of time to do so, ie let's force everything to use vanilla pytorch operations by having a default argument `--custom-cuda-kernels` and setting the value to `True` raises `NotImplementedError`.

Concerning the `loss_mask` yes, I'm pretty sure you need to mask out the the prefix part.

Concerning the position embeddings I'd say sequential, but let's get confirmation.

Concerning the choice of signature I think:
 - `prefix_len = None` is for no prefix-lm value. Though I know there's an enum so I have to figure that one out.
 - `prefix_len` is a double list otherwise:
     - first dimension is batch -> we can make an assert
     - second dimension is eod -> we can count that they correspond and make an assert
     - values are either int s.t `0 <= value < eod_index[j] - prev_index` or None (I don't like using negative values for the random behaviour). Perhaps it should even never be None, ie prefix are computed before hand as I'm sure we'll need the random values chosen for other processes.
     
If it's alright with you I can take over this (I'll look at your code in more details tmr), since I will be able to play on JZ with it.

**Comment by ontocord** [comment]
Yes please @thomasw21 take over. I'm happy to help however I can.

**Comment by ontocord** [comment]
@sbmaruf thanks for the links. I looked over the google doc and it looks like you all wanted to do some translation with mt5? I couldn't figure out from the original t5 code but admittedly I didn't spend long on it. Check out this comment though, which is interesting. https://github.com/tensorflow/mesh/blob/a4fb329e5a374e21e893906b178584baf237a9a6/mesh_tensorflow/transformer/transformer.py#L1027 

 In any case, my understanding is that we aren't doing the "masked span" model ala t5/mt5 for now (e.g., replacing spans of words of length 1, 2, 3, etc. with special tokens A <extra_id_0> C and predicting the missing tokens on the non-prefix side: <extra_id_0> B). But if we decide to go that route hopefully the code we are writing now will generalize as the "masking" is really just replacing some spans of tokens with a  special token and defining the right output sequence. 

Also for translation, assuming you can do something like "sentence in lang 1 <eos> translate to lang 2 <eos> sentence in lang 2 <eod>" where everything up to the second <eos> is the prefix?

**Comment by ontocord** [comment]
All, I added loss_on_targets_only parameter in the code above to mirror T5. This is set to true and defaults to not computing loss for the prefix portion.

**Comment by ontocord** [comment]
Also, @thomasw21 what do you think about the original way the function masks loss_mask. It assumes the whole sequence is filled in and that it's `Seq1 <eod> Seq2 <eod>... `. But if there is any junk token after any of the <eod>, it should mask the loss there too no? In particular, after the last `<eod>`?

**Comment by thomasw21** [comment]
What's a junk token? Maybe let's discuss during the meeting and post all the notes here afterwards?

**Comment by ontocord** [comment]
If all there are sequences in a batch that doesn't fill up the whole batch, i assume there's some token after the last eod, like <unk>. Do we really want to train on those tokens?

**Comment by ontocord** [comment]
This issue is being addressed in this PR: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/52

**Comment by ontocord** [comment]
This issue is being addressed in this PR: https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/52
